### PR TITLE
search: make search expr heuristic more strict to avoid confusion

### PR DESCRIPTION
**Pull request by rvantonder**
Fixes https://github.com/sourcegraph/sourcegraph/issues/34018.

This is a bit involved, I go into it in the issue comment above. But basically, the heuristic that decides how to parenthesize an unparenthesized ambiguous query will become more particular about when it applies, based on the order in which terms occur. You can read the issue description for details, but basically, before:

- We have a (desirable) heuristic that 
  - `repo:a b or d file:c` => 
  - `repo:a (b or d) file:c` => 
  - `repo:a file:c (b or d)`
  - Since this is a "best" guess according to tastes based on the original query

- Unfortunately, the same heuristic applies when the original ambiguous query is:
  - `repo:a b or file:c d` =>
  - `repo:a b or d file:c` (uh oh) =>
  - `repo:a (b or d) file:c` (as before) => 
  - `repo:a file:c (b or d)` D:

Because the heuristic doesn't care about the ordering of `repo` or `file` across the original parse tree `(repo:a b) or (file:c d)` when inspecting sub expressions. This PR fixes the "bad", second case, in general. To make the heuristic now care about ordering, I added a function tho report term ordering based on their position in the query string.

I still wish I could surface this in the frontend as "your query is interpreted as `such and such`". Maybe it's possible with an alert, may look into it later.

## Test plan
Updated tests.

### CODE REVIEW

File: internal/search/query/parser_test.go [hunk 1]
```diff
@@ -632,15 +632,15 @@ func TestParseAndOrLiteral(t *testing.T) {
 	autogold.Want("repo:foo (x", `(and "repo:foo" "(x") (HeuristicDanglingParens,Literal)`).Equal(t, test("repo:foo (x"))
 	autogold.Want("(x or bar() )", `(or "x" "bar()") (Literal)`).Equal(t, test("(x or bar() )"))
 	autogold.Want("(x", `"(x" (HeuristicDanglingParens,Literal)`).Equal(t, test("(x"))
-	autogold.Want("x or (x", `(or "x" "(x") (HeuristicDanglingParens,HeuristicHoisted,Literal)`).Equal(t, test("x or (x"))
```
Comment (rvantonder) [inline internal/search/query/parser_test.go:635]: Expected: the heuristic "fails" for the simple cases now, so doesn't get labeled but the interpretation is the same. We previously labeled this a bit overzealously.

File: internal/search/query/transformer_test.go [hunk 2]
```diff
@@ -137,32 +133,45 @@ func TestHoist(t *testing.T) {
...
 		},
-		// This next pattern is valid for the heuristic, even though the ordering of the
-		// patterns 'a' and 'c' in the first and last position are not ordered next to the
-		// 'or' keyword. This because no ordering is assumed for patterns vs. field:value
-		// parameters in the grammar. To preserve relative ordering and check this would
-		// impose significant complexity to PartitionParameters function during parsing, and
-		// the PartitionSearchPattern helper function that the heurstic relies on. So: we
-		// accept this heuristic behavior here.
 		{
-			input: `a repo:foo or b or file:bar c`,
```
Comment (rvantonder) [inline internal/search/query/transformer_test.go:151]: This is the buggy case that's now handled properly.

**Review by camdencheek — APPROVED** [review]
Nice. So, if I'm understanding this logic correctly, this no longer will incorrectly hoist a pattern group (the failing case), but it will also never hoist multiple pattern groups because it only does a single pass. For example, `repo:a b or c repo:b d or e` will not be interpreted as `repo:a (b or c) repo:b (d or e)`. Not a problem, just thinking out loud.

**Comment by rvantonder** [comment]
> but it will also never hoist multiple pattern groups because it only does a single pass

Yep! It's more strict now. I can add that as a test case.

**Comment by rvantonder** [comment]
welp, this is stuck, I can't seem to get a green build, but it seems to fail on building server image... can't tell anything from the `backend-integration` job. I'll need more time to investigate...

**Comment by rvantonder** [comment]
Hmmm maybe `main-dry-run` will success. let's seeee

**Pull request merged**